lib/quic.c

```diff
@@ -0,0 +1,983 @@
+                                     int *al, void *parse_arg)
+{
+  struct connectdata *conn = (struct connectdata *)SSL_get_app_data(ssl);
+  int rv;
+  ngtcp2_transport_params params;
+  (void)ext_type;
+  (void)context;
+  (void)x;
+  (void)chainidx;
+  (void)parse_arg;
+
+  rv = ngtcp2_decode_transport_params(
+    &params, NGTCP2_TRANSPORT_PARAMS_TYPE_ENCRYPTED_EXTENSIONS, in, inlen);
+  if(rv) {
+    fprintf(stderr, "ngtcp2_decode_transport_params: %s\n",
+            ngtcp2_strerror(rv));
+    *al = SSL_AD_ILLEGAL_PARAMETER;
+    return -1;
+  }
+
+  rv = ngtcp2_conn_set_remote_transport_params(
+    conn->quic.conn, NGTCP2_TRANSPORT_PARAMS_TYPE_ENCRYPTED_EXTENSIONS,
+    &params);
+  if(rv) {
+    *al = SSL_AD_ILLEGAL_PARAMETER;
+    return -1;
+  }
+
+  return 1;
+}
+
+static SSL_CTX *quic_ssl_ctx(struct Curl_easy *data)
+{
+  SSL_CTX *ssl_ctx = SSL_CTX_new(TLS_method());
+
+  SSL_CTX_set_min_proto_version(ssl_ctx, TLS1_3_VERSION);
+  SSL_CTX_set_max_proto_version(ssl_ctx, TLS1_3_VERSION);
+
+  /* This makes OpenSSL client not send CCS after an initial ClientHello. */
+  SSL_CTX_clear_options(ssl_ctx, SSL_OP_ENABLE_MIDDLEBOX_COMPAT);
+
+  SSL_CTX_set_default_verify_paths(ssl_ctx);
+
+  if(SSL_CTX_set_cipher_list(ssl_ctx, QUIC_CIPHERS) != 1) {
+    failf(data, "SSL_CTX_set_cipher_list: %s",
+          ERR_error_string(ERR_get_error(), NULL));
+    return NULL;
+  }
+
+  if(SSL_CTX_set1_groups_list(ssl_ctx, QUIC_GROUPS) != 1) {
+    failf(data, "SSL_CTX_set1_groups_list failed");
+    return NULL;
+  }
+
+  SSL_CTX_set_mode(ssl_ctx, SSL_MODE_QUIC_HACK);
+
+  if(SSL_CTX_add_custom_ext(ssl_ctx,
+                            NGTCP2_TLSEXT_QUIC_TRANSPORT_PARAMETERS,
+                            SSL_EXT_CLIENT_HELLO |
+                            SSL_EXT_TLS1_3_ENCRYPTED_EXTENSIONS,
+                            transport_params_add_cb,
+                            transport_params_free_cb, NULL,
+                            transport_params_parse_cb, NULL) != 1) {
+    failf(data, "SSL_CTX_add_custom_ext(NGTCP2_TLSEXT_QUIC_TRANSPORT_"
+          "PARAMETERS) failed: %s\n",
+          ERR_error_string(ERR_get_error(), NULL));
+    return NULL;
+  }
+
+  return ssl_ctx;
+}
+
+/** SSL callbacks ***/
+
+static void set_tls_alert(struct connectdata *conn,
+                          uint8_t alert)
+{
+  struct quicsocket *qs = &conn->quic;
+  qs->tls_alert = alert;
+}
+
+static int ssl_on_key(struct connectdata *conn,
+                      int name, const uint8_t *secret, size_t secretlen,
+                      const uint8_t *key, size_t keylen, const uint8_t *iv,
+                      size_t ivlen)
+{
+  int rv;
+  uint8_t pn[64];
+  ssize_t pnlen;
+  struct Context *crypto_ctx = &conn->quic.crypto_ctx;
+
+  switch(name) {
+  case SSL_KEY_CLIENT_EARLY_TRAFFIC:
+  case SSL_KEY_CLIENT_HANDSHAKE_TRAFFIC:
+  case SSL_KEY_CLIENT_APPLICATION_TRAFFIC:
+  case SSL_KEY_SERVER_HANDSHAKE_TRAFFIC:
+  case SSL_KEY_SERVER_APPLICATION_TRAFFIC:
```

Comment:
Error here also:
```
quic.c(307,8):  error: use of undeclared identifier 'SSL_KEY_CLIENT_EARLY_TRAFFIC'
  case SSL_KEY_CLIENT_EARLY_TRAFFIC:
       ^
quic.c(308,8):  error: use of undeclared identifier 'SSL_KEY_CLIENT_HANDSHAKE_TRAFFIC'
  case SSL_KEY_CLIENT_HANDSHAKE_TRAFFIC:
       ^
quic.c(309,8):  error: use of undeclared identifier 'SSL_KEY_CLIENT_APPLICATION_TRAFFIC'
  case SSL_KEY_CLIENT_APPLICATION_TRAFFIC:
       ^
quic.c(310,8):  error: use of undeclared identifier 'SSL_KEY_SERVER_HANDSHAKE_TRAFFIC'
  case SSL_KEY_SERVER_HANDSHAKE_TRAFFIC:
       ^
```

Reply:
Those are symbols that are added in ngtcp2's custom openssl version. Did you build and install that and made sure to point out that build when you build this curl? I mention this in the docs/HTTP3.md file but I figure it is easily missed!

Reply:
>  Did you build and install that ..

No, I missed that. 

BTW. it seems ngtcp2 is not yet quite Windows-ready. Includes `<unistd.h>`, but not `<winsock2.h>` etc.

Reply:
I now d/l and built with this `https://github.com/tatsuhiro-t/openssl/`. The `SSL_MODE_QUIC_HACK` etc. is nowhere to be found there.  Are you sure your WIP is up-to-date against Tatsuhiro's OpenSSL?

Reply:
That define is used by ngtcp2's own client.cc example:

https://github.com/ngtcp2/ngtcp2/blob/c0cd76426965fd7162132523a9dd04e6a2f9c200/examples/client.cc#L1941

... even in their draft-15 branch which I believe is where they're working mostly right now.

I'm using the quic-draft-14 branch of @tatsuhiro-t's openssl (as the ngtcp2 README says so) and that defines `SSL_MODE_QUIC_HACK` here:

https://github.com/tatsuhiro-t/openssl/blob/bfd4950424bfc1508f1e70cc0ca9de772ed38b11/include/openssl/ssl.h#L497

> it seems ngtcp2 is not yet quite Windows-ready

I'm sure they'll appreciate a pull-request that remedies that! =)